### PR TITLE
[staking] handle typedCandidate registration

### PR DESCRIPTION
**Pull request by envestcc**
# Description
The pr introduced the `TypedCandidate`, and implemented its general registration process framework, laying the foundation for adding specific type node registrations in the future.

Fixes #(issue)

## Type of change
Please delete options that are not relevant.
- [] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [] Code refactor or improvement
- [] Breaking change (fix or feature that would cause a new or changed behavior of existing functionality)
- [] This change requires a documentation update

# How Has This Been Tested?
Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration
- [x] make test
- [] fullsync
- [] Other test (please specify)

**Test Configuration**:
- Firmware version:
- Hardware:
- Toolchain:
- SDK:

# Checklist:
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: action/protocol/staking/typedcandidate.go [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+		Operator           address.Address
+		Reward             address.Address
+		Type               CandidateType
+		SelfStakeBucketIdx uint64
+		// extraData is the customised data for the candidate
+		extraData []byte
+	}
+)
+
+func candidateExtra[T interface{ Deserilized([]byte) error }](c *TypedCandidate, extra T) error {
```
Comment (dustinxie) [inline action/protocol/staking/typedcandidate.go:43]: `Deserialize()`

File: action/protocol/staking/handlers.go [hunk 2]
```diff
@@ -775,6 +776,83 @@ func (p *Protocol) handleCandidateUpdate(ctx context.Context, act *action.Candid
 	return log, nil
 }
 
+func (p *Protocol) handleTypedCandidateRegister(ctx context.Context, act typedCandidateRegister, csm CandidateStateManager, esm *typedCandidateStateManager,
```
Comment (Liuhaai) [inline action/protocol/staking/handlers.go:779]: where is `handleTypedCandidateRegister` called?

File: action/protocol/staking/typedcandidate.go [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+package staking
+
+import (
+	"math/big"
+
+	"github.com/iotexproject/iotex-address/address"
+)
+
+type (
+	// typedCandidateRegister is the interface that every candidate register action should implement
```
Comment (Liuhaai) [inline action/protocol/staking/typedcandidate.go:15]: private or public?

File: action/protocol/staking/handlers.go [hunk 4]
```diff
@@ -775,6 +776,83 @@ func (p *Protocol) handleCandidateUpdate(ctx context.Context, act *action.Candid
...
+	registrationFee := new(big.Int).Set(p.config.RegistrationConsts.Fee)
+
+	caller, fetchErr := fetchCaller(ctx, esm, new(big.Int).Add(actAmount, registrationFee))
+	if fetchErr != nil {
+		return log, nil, fetchErr
+	}
+	owner := actCtx.Caller
+	if act.OwnerAddress() != nil {
+		owner = act.OwnerAddress()
+	}
```
Comment (Liuhaai) [inline action/protocol/staking/handlers.go:796]: else {owner := actCtx.Caller}

File: action/protocol/staking/handlers.go (hunk 2 above)
Comment (Liuhaai) [reply action/protocol/staking/handlers.go:779]: Can similar code be shared with `handleCandidateRegister`

File: action/protocol/staking/typedcandidate.go (hunk 1 above)
Comment (Liuhaai) [reply action/protocol/staking/typedcandidate.go:43]: extra func in this pr

File: action/protocol/staking/handlers.go [hunk 5]
```diff
@@ -775,6 +776,83 @@ func (p *Protocol) handleCandidateUpdate(ctx context.Context, act *action.Candid
...
+
+	// cannot collide with existing operator at same type
+	if esm.isRegistered(candType, act.OperatorAddress()) {
+		return log, nil, &handleError{
+			err:           ErrInvalidOperator,
+			failureStatus: iotextypes.ReceiptStatus_ErrCandidateConflict,
+		}
+	}
+
+	bucket := NewVoteBucket(owner, owner, actAmount, act.Duration(), blkCtx.BlockTimeStamp, act.AutoStake())
```
[the remark this answers is not in the document]
Comment (Liuhaai) [reply action/protocol/staking/handlers.go:806]: should amount be verified?

File: action/protocol/staking/handlers.go (hunk 2 above)
Comment (envestcc) [reply action/protocol/staking/handlers.go:779]: `handleTypedCandidateRegister` will be called when introducing a specific register action

Comment (envestcc) [reply action/protocol/staking/handlers.go:779]: `handleTypedCandidateRegister` uses different candidate state storage than `handleCandidateRegister`

File: action/protocol/staking/typedcandidate.go (hunk 3 above)
Comment (envestcc) [reply action/protocol/staking/typedcandidate.go:15]: it's a interface used within this package, so should be private

File: action/protocol/staking/handlers.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (envestcc) [reply action/protocol/staking/handlers.go:806]: > separate staking for multi roles?

yes

[the remark this answers is not in the document]
Comment (envestcc) [reply action/protocol/staking/handlers.go:806]: the validation should be implemented in the `Validate` of `staking.Protocol` by specific register action, just like other actions do

File: action/protocol/staking/handlers.go (hunk 2 above)
Comment (Liuhaai) [reply action/protocol/staking/handlers.go:779]: > `handleTypedCandidateRegister` will be called when introducing a specific register action

Add TODO?

**Review by Liuhaai — APPROVED** [review]

**Review by bitcoinfinancier — APPROVED** [review]

**Comment by envestcc** [comment]
IIP-16 paused

**Pull request closed without merging**